cache/network/api.go

```diff
@@ -0,0 +1,230 @@
+package network
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+const maxKeyLength = 512
+const maxKeyCount = 8
+
+type prepareUploadRequest struct {
+	CacheKey           string `json:"cache_key"`
+	ArchiveFileName    string `json:"archive_filename"`
+	ArchiveContentType string `json:"archive_content_type"`
+	ArchiveSizeInBytes int64  `json:"archive_size_in_bytes"`
+}
+
+type prepareUploadResponse struct {
+	ID            string            `json:"id"`
+	UploadMethod  string            `json:"method"`
+	UploadURL     string            `json:"url"`
+	UploadHeaders map[string]string `json:"headers"`
+}
+
+type restoreResponse struct {
+	URL string `json:"url"`
+}
+
+type apiClient struct {
+	httpClient  *retryablehttp.Client
+	baseURL     string
+	accessToken string
+}
+
+func newAPIClient(client *retryablehttp.Client, baseURL string, accessToken string) apiClient {
+	return apiClient{
+		httpClient:  client,
+		baseURL:     baseURL,
+		accessToken: accessToken,
+	}
+}
+
+func (c apiClient) prepareUpload(requestBody prepareUploadRequest) (prepareUploadResponse, error) {
+	url := fmt.Sprintf("%s/upload", c.baseURL)
```

Comment:
NOTE: you could work with url.URL type for a bit nicer URL handling.

Storing the base URL on the API client as `*url.URL`:

```
	baseURL, err := url.Parse(baseURL)
	if err != nil {
		panic("invalid api base url: " + err.Error())
	}
```

and adding the endpoint to the base URL with `url.Parse`:

```
	u, err := c.BaseURL.Parse(endpoint)
	if err != nil {
		return nil, fmt.Errorf("parsing endpoint failed: %v", err)
	}
```

to execute the request you can get the raw URL with `u.String()`

Reply:
I don't know, it feels like a lot of boilerplate and extra error handling for a simple thing.

I did notice though that Go 1.19 adds `url.JoinPath()`, which will be much simpler: https://pkg.go.dev/net/url#JoinPath